File: removeDuplicate.cpp
#include <vector>
#include <iostream>

int removeDuplicates(std::vector<int>& nums) {
    for (auto itor = nums.begin(); itor != nums.end(); itor++){
        if(itor == nums.begin()){
            continue;
        }
        if(*itor == *(itor - 1)){
            itor = nums.erase(itor-1);
        }
    }
    return nums.size();
}

// Before = {1,1,1,2,2,3}
// After  = {1,2,3}

// Before = {1,1}
// After  = {1}

// Before = {1,2,3}
// After  = {1,2,3}

// Before = {0,0,1,1,1,1,2,3,3}
// After  = {0,1,2,3}

int main(){
    std::vector<std::vector<int>> testSet = {
        {1,1,1,2,2,3}, {1,1},{1,2,3},{0,0,1,1,1,1,2,3,3}
    };

    for(std::vector<int> test : testSet){
        std::cout << "Before = {";
        for(int i = 0; i < test.size(); i++){
            std::cout << test[i] << (i == (test.size() - 1)? "}\n":",");
        }
        removeDuplicates(test);
        std::cout << "After  = {";
        for(int i = 0; i < test.size(); i++){
            std::cout << test[i]  << (i == (test.size() - 1)? "}\n\n":",");
        }
    }

    return 0;
}

File: Java/easy/removeDuplicate.java
import java.util.*;

class Solution {
    public static int removeDuplicates(int[] nums) {
        if(nums.length == 1) return 1;
        int fast = 1, slow = 0;
        int n = nums.length;
        while(fast < n){
            if(nums[fast] == nums[slow]){
                fast++;
            }
            else{
                slow++;
                nums[slow] = nums[fast];
            }
        }

        return slow + 1;
    }

    // Original Array = [1, 1, 2]      After removing duplicates = [1,2]
    // Original Array = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4] After removing duplicates = [0,1,2,3,4]

    public static void main(String[] args){
        ArrayList<int[]> testSet = new ArrayList<int[]>();
        testSet.add(new int[]{1,1,2});
        testSet.add(new int[]{0,0,1,1,1,2,2,3,3,4});

        for(int[] test : testSet){
            System.out.print("Original Array = " + Arrays.toString(test) + "\tAfter removing duplicates = [");
            int stop = removeDuplicates(test);
            for(int i = 0; i < stop; i++){
                System.out.print(test[i] + (i == (stop - 1)?"]\n":","));
            }
        }
    }
}